File: my-folder/2903-insert-greatest-common-divisors-in-linked-list/solution.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {ListNode}
def insert_greatest_common_divisors(head)
  cur = head
  while cur.next
    insert = ListNode.new(cur.val.gcd(cur.next.val), cur.next)
    cur.next = insert
    cur = insert.next
  end
  head
end

File: my-folder/2903-insert-greatest-common-divisors-in-linked-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def gcd(a, b):
          if a == 0:
            return b

          return gcd(b % a, a)
        
        curr = head

        while curr.next:
          nxt = curr.next
          # find gcd of curr.val and curr.next -> temp
          temp = ListNode(gcd(curr.val, nxt.val), nxt)
          curr.next = temp
          curr = nxt

        return head
    





        
